File: NinePSharp.Messages.FSharp/messages/Rmknod.fs
namespace NinePSharp.Messages.FSharp

open System
open System.Buffers.Binary
open NinePSharp.Constants
open NinePSharp.Interfaces
open NinePSharp.Messages
open NinePSharp.Messages.FSharp.Protocol

type Rmknod =
    { Size: uint32
      Tag: uint16
      Qid: Qid }

    interface ISerializable with
        member this.Size = this.Size
        member this.Type = MessageTypes.Rmknod
        member this.Tag = this.Tag
        member this.WriteTo(span: Span<byte>) =
            let mutable offset = 0
            writeHeaders this.Size this.Tag MessageTypes.Rmknod span &offset
            writeQid this.Qid span &offset

    static member Parse (data: ReadOnlySpan<byte>) =
        let mutable offset = 0
        let size, tag = readHeaders data &offset
        let qid = readQid data &offset
        { Size = size; Tag = tag; Qid = qid }

File: NinePSharp/messages/Rmknod.cs
namespace NinePSharp.Messages;

using System;
using System.Buffers.Binary;
using NinePSharp.Constants;
using NinePSharp.Interfaces;
using NinePSharp.Protocol;

public readonly struct Rmknod : ISerializable
{
    public uint Size { get; }
    public MessageTypes Type => MessageTypes.Rmknod;
    public ushort Tag { get; }
    public Qid Qid { get; }

    public Rmknod(uint size, ushort tag, Qid qid)
    {
        Size = size;
        Tag = tag;
        Qid = qid;
    }

    public Rmknod(ReadOnlySpan<byte> data)
    {
        Size = BinaryPrimitives.ReadUInt32LittleEndian(data[..4]);
        Tag = BinaryPrimitives.ReadUInt16LittleEndian(data.Slice(5, 2));
        int offset = NinePConstants.HeaderSize;
        Qid = data.ReadQid(ref offset);
    }

    public void WriteTo(Span<byte> span)
    {
        span.WriteHeaders(Size, Tag, MessageTypes.Rmknod);
        int offset = NinePConstants.HeaderSize;
        span.WriteQid(Qid, ref offset);
    }
}
